python/dgl/transform.py

```diff
@@ -714,6 +717,131 @@ def compact_graphs(graphs, always_preserve=None):
+        Whether to always include the right hand side nodes in the left hand side.
+    lhs_suffix : str, default "_l"
+        The suffix attached to all node types on the left hand side.
+    rhs_suffix : str, default "_r"
+        The suffix attached to all node types on the right hand side.
+
+    Returns
+    -------
+    DGLHeteroGraph
+        The new graph with the bipartite structure.
+
+        The node IDs induced for each type in both sides would be stored in feature
+        ``dgl.NID``.
+
+        The edge IDs induced for each type would be stored in feature ``dgl.EID``.
+
+        If ``include_rhs_in_lhs`` is True, then for each node type ``ntype``, the first
+        few nodes with type ``ntype_l`` are guaranteed to be identical to the nodes with
+        type ``ntype_r`` (in the sense of both values and order).
+
+    Notes
+    -----
+    A special use case of this function is to create a graph structure for efficient
+    computation of message passing.  See [TODO] for a detailed example.
+    """
+    if rhs_nodes is None:
+        # Find all nodes that appeared as destinations
+        rhs_nodes = defaultdict(list)
+        for etype in g.canonical_etypes:
+            _, dst = g.edges(etype=etype)
+            rhs_nodes[etype[2]].append(dst)
+        rhs_nodes = {ntype: F.unique(F.cat(values, 0)) for ntype, values in rhs_nodes.items()}
+    elif not isinstance(rhs_nodes, Mapping):
+        # rhs_nodes is a Tensor, check if the g has only one type.
+        if len(g.ntypes) > 1:
+            raise ValueError(
+                'Graph has more than one node type; please specify a dict for rhs_nodes.')
+        rhs_nodes = {g.ntypes[0]: rhs_nodes}
+
+    # rhs_nodes is now a dict
+    rhs_nodes = [rhs_nodes.get(ntype, F.tensor([], dtype=F.int64)) for ntype in g.ntypes]
+    rhs_nodes_nd = [F.zerocopy_to_dgl_ndarray(nodes) for nodes in rhs_nodes]
+
+    new_graph_index, lhs_nodes_nd, induced_edges_nd = _CAPI_DGLToBipartite(
+        g._graph, rhs_nodes_nd, include_rhs_in_lhs)
+    lhs_nodes = [F.zerocopy_from_dgl_ndarray(nodes_nd.data) for nodes_nd in lhs_nodes_nd]
+
+    new_ntypes = [ntype + lhs_suffix for ntype in g.ntypes] + \
+                 [ntype + rhs_suffix for ntype in g.ntypes]
+    new_graph = DGLHeteroGraph(new_graph_index, new_ntypes, g.etypes)
+
+    for i, ntype in enumerate(g.ntypes):
+        new_graph.nodes[ntype + lhs_suffix].data[NID] = lhs_nodes[i]
+        new_graph.nodes[ntype + rhs_suffix].data[NID] = rhs_nodes[i]
+
+    for i, canonical_etype in enumerate(g.canonical_etypes):
+        induced_edges = F.zerocopy_from_dgl_ndarray(induced_edges_nd[i].data)
+        utype, etype, vtype = canonical_etype
+        new_canonical_etype = (utype + lhs_suffix, etype, vtype + rhs_suffix)
+        new_graph.edges[new_canonical_etype].data[EID] = induced_edges
+
+    return new_graph
+
+def remove_edges(g, edge_ids):
+    """Return a new graph with given edge IDs removed.
```

Comment:
Does it relabel node ids?

Reply:
No, the nodes are preserved; I should mention it.